### ISSUE TITLE
Non-arithmetic operator precedence

### ISSUE THREAD

**Issue by yitzhaks** [DotNetAnalyzers/StyleCopAnalyzers#3167]
I'm aware of SA1407ArithmeticExpressionsMustDeclarePrecedence which is intended to improve the maintainability of arithmetic expressions.

I couldn't however find a rule that will flag the following:
```
bool? a = true;
if (a ?? false && b) ...
```
The problem is that since operator && takes precedence over operator ?? (according to the [documentation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/)), the above is equivalent to:
```
if (a ?? (false && b)) => if (a ?? false)
```
I assume no reasonable developer would write the original code with the intent of optimizing away his second condition at compile-time, but that's what happens, and it isn't flagged neither by the compiler nor by StyleCop.Analyzers.
I found this out the hard way when debugging some unexpected behavior in my code :)

In the hopes of making the world a better place, I'd like to propose either than SA1407 be expanded, or that a new maintainability rule be added for non-arithmetic operator precedence (I'm not sure how to generalize the rule for operators other than null-coalescing and conditional, but perhaps someone else has a more general form for the rule).

**Comment by sharwell** [comment]
I believe this is closer to [SA1408](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1408.md).

I'd be interested to know how many times this syntax occurs in real-world projects. I'm also interested to see whether Visual Studio's "unnecessary parentheses" feature flags this case.

**Comment by yitzhaks** [comment]
@sharwell, thanks. I agree it's closer to SA1408.
I don't follow your comment about "unnecessary parentheses", unless you mean to ask whether Visual Studio would have flagged the parentheses if they were added.

**Comment by sharwell** [comment]
> unless you mean to ask whether Visual Studio would have flagged the parentheses if they were added.

Yes, I'm referring to this.

**Comment by yitzhaks** [comment]
> I'd be interested to know how many times this syntax occurs in real-world projects.

When I found this was my problem, I easily found one more example of it in our codebase. I think it's not rare at all. The scenario is:
```
// if myObject is null we shouldn't enter this if, otherwise, check SomeValue and otherCondition
if (myObject?.SomeValue ?? false && otherCondition)
{
    ...
}
```

**Comment by bjornhellander** [comment]
@sharwell, regarding your question about if VS flags parenthesis as unnecessary, neither of these cases are flagged by VS:
```
(a ?? false) && b
a ?? (false && b)
```

Looks like an easy mistake to make, so I would also very much like this enhancement to e.g. SA1408.
